workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.test.ts

```diff
@@ -0,0 +1,347 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  IdentityPermission,
+  DynamicPermissionsPlugin,
+  DynamicAuthorizationService,
+  DynamicOperation,
+  ForbiddenError,
+  AuthorizationPlugin,
+  AuthenticatedUser
+} from '../';
+
+describe('Dynamic Authorization Service', () => {
+  let dynamicAuthorizationService: DynamicAuthorizationService;
+  let mockAuthorizationPlugin: AuthorizationPlugin;
+  let mockDynamicPermissionsPlugin: DynamicPermissionsPlugin;
+  beforeEach(() => {
+    jest.clearAllMocks();
+    mockAuthorizationPlugin = {
+      isAuthorized: jest.fn(),
+      isAuthorizedOnDynamicOperations: jest.fn()
+    };
+    mockDynamicPermissionsPlugin = {
+      createGroup: jest.fn(),
+      deleteGroup: jest.fn(),
+      getUserGroups: jest.fn(),
+      getUsersFromGroup: jest.fn(),
+      assignUserToGroup: jest.fn(),
+      removeUserFromGroup: jest.fn(),
+      createIdentityPermissions: jest.fn(),
+      deleteIdentityPermissions: jest.fn(),
+      deleteSubjectPermissions: jest.fn(),
+      getIdentityPermissionsByIdentity: jest.fn(),
+      getIdentityPermissionsBySubject: jest.fn()
+    };
+    dynamicAuthorizationService = new DynamicAuthorizationService({
+      dynamicPermissionsPlugin: mockDynamicPermissionsPlugin,
+      authorizationPlugin: mockAuthorizationPlugin
+    });
+  });
+
+  describe('isAuthorizedOnSubject', () => {
+    const mockAuthenticatedUser: AuthenticatedUser = {
+      id: 'sampleId',
+      roles: ['groupId0']
+    };
+    const groupIds = ['groupId1'];
+    const sampleGroupPermissions: IdentityPermission[] = [
+      {
+        effect: 'ALLOW',
+        action: 'READ',
+        subjectType: 'sampleSubject',
+        subjectId: 'sampleSubjectId',
+        identityType: 'GROUP',
+        identityId: 'groupId1'
+      }
+    ];
+    const sampleGroupWildcardPermissions: IdentityPermission[] = [
+      {
+        effect: 'ALLOW',
+        action: 'READ',
+        subjectType: 'sampleSubject',
+        subjectId: '*',
+        identityType: 'GROUP',
+        identityId: 'groupId1'
+      }
+    ];
+
+    const mockDynamicOperation: DynamicOperation = {
+      action: 'READ',
+      subjectType: 'sampleSubject',
+      subjectId: 'sampleSubjectId'
+    };
+    beforeEach(() => {
+      jest.spyOn(mockDynamicPermissionsPlugin, 'getUserGroups').mockResolvedValueOnce({
+        groupIds
+      });
+
+      jest.spyOn(mockDynamicPermissionsPlugin, 'getIdentityPermissionsBySubject').mockResolvedValueOnce({
+        identityPermissions: sampleGroupPermissions
+      });
+
+      jest.spyOn(mockDynamicPermissionsPlugin, 'getIdentityPermissionsBySubject').mockResolvedValueOnce({
+        identityPermissions: sampleGroupWildcardPermissions
+      });
+    });
+
+    test('Check user with a valid permissions for a dynamic operation', async () => {
```

Comment:
this test has lots of different `expects` in it. We should try to keep one expect per test to ensure we remove situations where tests end up having hidden dependencies that are needed for it to pass.